File: leetsync/2473-max-sum-of-a-pair-with-equal-sum-of-digits/max-sum-of-a-pair-with-equal-sum-of-digits.go
func maximumSum(nums []int) int {
	digitSumToNumMap := [82]int{}
	maxSum := -1
	for _, num := range nums {
		digitSum := getDigitSum(num)
		value := digitSumToNumMap[digitSum]
		if value > 0 {
			maxSum = max(maxSum, num+value)
		} else {
			value = 0
		}
		digitSumToNumMap[digitSum] = max(num, value)
	}
	return maxSum
}

func getDigitSum(num int) int {
	sum := 0
	for num > 0 {
		sum = sum + num%10
		num = num / 10
	}
	return sum
}

File: leetsync/2473-max-sum-of-a-pair-with-equal-sum-of-digits/max-sum-of-a-pair-with-equal-sum-of-digits.java
class Solution {
    public int maximumSum(int[] nums) {
        int n = nums.length;
        Map<Integer, Integer> digitSumToNumMap = new HashMap<>();
        int maxSum = -1;
        for (int num : nums) {
            int digitSum = getDigitSum(num);
            if (digitSumToNumMap.containsKey(digitSum)) {
                maxSum = Math.max(maxSum, num + digitSumToNumMap.get(digitSum));
            }
            digitSumToNumMap.put(digitSum, Math.max(num, digitSumToNumMap.getOrDefault(digitSum, 0)));
        }
        return maxSum;
    }

    public int getDigitSum(int num) {
        int sum = 0;
        while (num > 0) {
            sum += num % 10;
            num = num / 10;
        }
        return sum;
    }
}